add functions to random matrix generator

alfrisch opened the pull request:
### Summary

This PR adds functions random_hermitian() and random_nonhermitian() to random_matrix_generator.py in Aqua utilities folder.

### Details and comments

Several properties for generated matrices can be specified, like eigenvalues, eigenvalue range, condition number, sparsity, and truncation of Pauli matrices.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/aqua?pullRequest=192) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/aqua?pullRequest=192) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: alfrisch<br/>:x: dbucher97<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/aqua?pullRequest=192) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/aqua?pullRequest=192) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/aqua?pullRequest=192) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: alfrisch<br/>:x: dbucher97<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/aqua?pullRequest=192) it.</sub>

chunfuchen commented:
@dbucher97 do you mind signing CLA?

alfrisch commented:
@chunfuchen regarding the CLA there are two entries by the CLAassistant, the first one shows CLA signed (including @dbucher97) the second entry shows the missing signature, hitting "recheck" does not update the second entry. Something is strange here.
In the PR requirements it is reporting that all checks have passed, including CLA.

chunfuchen commented:
@alfrisch thanks, I was afraid the CLA will block the merge but it seems not.

hushaohan commented:
@alfrisch I'm not sure if the output below is expected:

```
>>> from random_matrix_generator import *
>>> random_unitary(1)
array([[0.+0.j]])
>>> a = random_unitary(2)
>>> a
array([[ 0.00000000e+00+0.70710678j, -1.38777878e-17+0.1767767j ],
       [ 0.00000000e+00+0.70710678j,  8.32667268e-17-0.1767767j ]])
>>> a.dot(a.T.conj())
array([[0.53125+0.00000000e+00j, 0.46875+1.22663473e-17j],
       [0.46875-1.22663473e-17j, 0.53125+0.00000000e+00j]])
```

alfrisch commented:
@hushaohan, yes, this is not what we expect for unitary matrices! This code was outside of the scope of this PR changes and thus was not tested. But anyway I could find the bug in random_unitary() and fixed it.

chunfuchen commented:
@alfrisch if possible, do yo mind adding one or two simple tests for this function? Thanks.

alfrisch commented:
@chunfuchen I added two unittests for random_unitary() and random_hermitian() for tiny, medium and large matrix dimensions
@hushaohan thanks for pointing that out, switching to the divide and having random integers between 0 and 2 caused in some cases the nan values, I also added the special case for N=1

hushaohan commented:
@alfrisch the `N=1` special case handling doesn't produce unit vectors; Getting rid of it (i.e., keeping only the code under `else`) actually works fine for `N=1,2,3...`.

alfrisch commented:
@hushaohan I changed that. It previously did not work with integer random numbers, but it works fine with floats, as you pointed out.
